connectors/src/connectors/confluence/lib/cli.ts

```diff
@@ -0,0 +1,247 @@
+import type {
+  AdminSuccessResponseType,
+  ConfluenceCommandType,
+  ConfluenceUpsertPageResponseType,
+  ModelId,
+} from "@dust-tt/types";
+import assert from "assert";
+import fs from "fs/promises";
+
+import {
+  getConfluencePageParentIds,
+  getSpaceHierarchy,
+} from "@connectors/connectors/confluence/lib/hierarchy";
+import {
+  confluenceGetSpaceNameActivity,
+  fetchConfluenceConfigurationActivity,
+  getConfluenceClient,
+  upsertConfluencePageInDb,
+  upsertConfluencePageToDataSource,
+} from "@connectors/connectors/confluence/temporal/activities";
+import { dataSourceConfigFromConnector } from "@connectors/lib/api/data_source_config";
+import { ConfluencePage } from "@connectors/lib/models/confluence";
+import { default as topLogger } from "@connectors/logger/logger";
+import { ConnectorResource } from "@connectors/resources/connector_resource";
+
+interface cachedSpace {
+  spaceName: string | null;
+  spaceHierarchy: Record<string, string | null>;
+}
+
+async function cacheSpace(
+  cachedSpaces: Record<string, cachedSpace>,
+  {
+    confluenceCloudId,
+    connectorId,
+    spaceId,
+  }: {
+    confluenceCloudId: string;
+    connectorId: ModelId;
+    spaceId: string;
+  }
+): Promise<cachedSpace> {
+  const cachedSpace = cachedSpaces[spaceId];
+  if (cachedSpace) {
+    return cachedSpace;
+  }
+  const spaceName = await confluenceGetSpaceNameActivity({
+    spaceId,
+    confluenceCloudId,
+    connectorId,
+  });
+  const spaceHierarchy = spaceName
+    ? await getSpaceHierarchy(connectorId, spaceId)
+    : {}; // not fetching if we couldn't get the space from Confluence API anyway
+  cachedSpaces[spaceId] = { spaceName, spaceHierarchy };
+  return { spaceName, spaceHierarchy };
+}
+
+export const confluence = async ({
+  command,
+  args,
+}: ConfluenceCommandType): Promise<
+  AdminSuccessResponseType | ConfluenceUpsertPageResponseType
+> => {
+  const logger = topLogger.child({ majorCommand: "confluence", command, args });
+  switch (command) {
+    case "upsert-page": {
```

Comment:
There's a bit of a risk of having some divergence between the cli code and the activities code, we generally try to call an activity from here even if that means it's async.

All that being said I'm fine with having a direct implementation. Any chance we can dry a bit these 2 commands?

Reply:
yep I agree, I was contemplating adding a new activity but in the end we won't do that in the workflow (upsert and check parents right after)

Reply:
Could be an option on the activity + a new workflow to do just one page with that option set

Reply:
hmm interesting

Reply:
sounds good, doing that 👍

Reply:
actually it's a bit too convoluted to use the same activity: one fetches the page + upserts without parent and is optimized for the workflow (it does some version checking), the other one fetches the page, fetches the space using the page data, resolves parents and does an upsert with parents.

I added a separate activity to do the `upsertWithFullParents` and added workflows since we might want to do these with retries and history

Reply:
Yess!